carabiner-to-py/CarabinerToPy.py

```diff
@@ -0,0 +1,80 @@
+import socket
+import edn_format
+import time
+
+class LinkListener():
+    """A simple python client to communicate with carabiner (a Abelton Link connector)
+    Carabiner server must be running to use. Requires edn_format [$pip install edn_format]"""
+    def __init__(self, tcp_ip = '127.0.0.1', tcp_port = 17000, buffer_size = 1024):
+        self.tcp_ip = tcp_ip
+        self.tcp_port = tcp_port
+        self.buffer_size = buffer_size
+        self.s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.s.connect((tcp_ip, tcp_port))
+
+    def decode_edn_msg(self, msg):
+        """Decodes a TCP message from Carabiner to python dictionary"""
+        msg = msg.decode()
+        striped_msg = msg[msg.index('{'):]
+        decoded_msg = edn_format.loads(striped_msg)
+
+        # Because the edf_format package does not return normal dam dicts (or string keywords). What dicts.
+        if type(decoded_msg) is edn_format.immutable_dict.ImmutableDict:
+            decoded_dict = {}
+            for key, value in decoded_msg.dict.items():
+                print()
+                decoded_dict[str(key).strip(':')] = value
+            decoded_msg = decoded_dict
+
+        return decoded_msg
+
+    def status(self):
+        """Wrapper for Status"""
+        self.s.send(b'status')
+        return_msg = self.s.recv(self.buffer_size)
+        return self.decode_edn_msg(return_msg)

```

Comment:
This looks like a nice start for a Python interface to Carabiner. I don’t know Python well enough to judge how idiomatic or efficient it is, but a lot of the right pieces seem to be there. The biggest thing that I see missing is that you need a background thread to read messages from Carabiner to cope with the fact that you will receive them asynchronously, whenever the Link session tempo or start point changes, or peers come and go. So you will probably want to register a callback function in Python that will be called to update the client code when things like that happen. You can find my Clojure implementation of that thread for beat-link-trigger [here](https://github.com/brunchboy/beat-link-trigger/blob/master/src/beat_link_trigger/carabiner.clj#L148-L187), in case that is any help in designing a Python equivalent. The rest of the file might be interesting too, it is where I have all the Carabiner communication for that program.

I do think that by the time this becomes a complete implementation (and you can add `request_beat_at_time`, and finish off `time_at_beat` and `phase_at_time` because I have added those to Carabiner for you), it will be large enough to merit being its own project, with its own documentation, and I will happily point at it from the Carabiner documentation.